### PR TITLE
Facebook Seq2seq model in cocoa framework

### PR DESCRIPTION
**Pull request by derekchen14**

### CODE REVIEW

File: cocoa/core/dataset.py [hunk 1]
```diff
@@ -114,10 +114,10 @@ def read_examples(paths, max_examples, Scenario):
     examples = []
     for path in paths:
         print 'read_examples: %s' % path
-        for raw in read_json(path):
+        for example in read_json(path):
             if max_examples >= 0 and len(examples) >= max_examples:
                 break
-            examples.append(Example.from_dict(raw, Scenario))
+            examples.append(Example.from_dict(example, Scenario))
     return examples
```
Comment (hhexiy) [inline cocoa/core/dataset.py]: hmm. not sure why you changed raw to example. but raw is more informative here. example is like the Example object.

File: cocoa/neural/evaluator.py [hunk 2]
```diff
@@ -82,7 +82,7 @@ def evaluate(self, opt, model_opt, data, split='test'):
                 gold_words_total += len(response.gold_sent)
 
             if opt.verbose:
-                self.print_results(model_opt, batch, utterances)
+                counter = self.print_results(model_opt, batch, counter, utterances)
```
Comment (hhexiy) [inline cocoa/neural/evaluator.py:80]: what's the counter for?

File: cocoa/neural/generator.py [hunk 3]
```diff
@@ -78,8 +75,9 @@ def _run_attention_memory(self, batch, enc_memory_bank):
                 memory_bank.append(desc_memory_bank)
             elif hasattr(batch, 'scene_inputs'):
```
Comment (hhexiy) [inline cocoa/neural/generator.py]: this should be in the inherited class in fb_negotiation.

File: cocoa/neural/loss.py [hunk 4]
```diff
@@ -48,7 +48,7 @@ def compute_loss(self, target, output):
         # reward: (batch_size,)
         batch_size = output.size(1)
         #aeq(batch_size, reward.size(0))
-
+        print("comes to reinforce")
```
Comment (hhexiy) [inline cocoa/neural/loss.py]: clean up debug print

File: cocoa/neural/models.py [hunk 5]
```diff
@@ -81,7 +81,8 @@ class MeanEncoder(EncoderBase):
     """
     def __init__(self, num_layers, embeddings, embed_type='utterance'):
         super(MeanEncoder, self).__init__()
-        self.num_layers = num_layers
+        n_layers = 1 if embed_type == "scenario" else num_layers
+        self.num_layers = n_layers
```
Comment (hhexiy) [inline cocoa/neural/models.py]: num_layers should be specified in cmd line. should not be changed here in an ad-hoc way.

File: cocoa/neural/models.py [hunk 6]
```diff
@@ -91,9 +92,8 @@ def forward(self, src, lengths=None, encoder_state=None):
 
         emb = self.embeddings(src)
         s_len, batch, emb_dim = emb.size()
-        mean = emb.mean(0).expand(self.num_layers, batch, emb_dim)
+        encoder_final = emb.mean(0).expand(self.num_layers, batch, emb_dim)
         memory_bank = emb
```
Comment (hhexiy) [inline cocoa/neural/models.py]: why do you change encoder_final here?

File: cocoa/neural/trainer.py [hunk 7]
```diff
@@ -238,7 +238,14 @@ def validate(self, valid_iter):
...
             outputs, attns, _ = self._run_batch(batch, None, enc_state)
-            _, batch_stats = self.valid_loss.compute_loss(batch.targets, outputs)
+
+            # TODO: clean up this check
+            if hasattr(batch, "selections"):
+                _, batch_stats = self.valid_loss.compute_loss(batch.targets,
+                                    batch.selections, outputs)
+            else:
+                _, batch_stats = self.valid_loss.compute_loss(batch.targets, outputs)
+
```
Comment (hhexiy) [inline cocoa/neural/trainer.py]: put this in fb trainer

File: cocoa/neural/trainer.py [hunk 8]
```diff
@@ -322,14 +329,23 @@ def _run_seq2seq_batch(self, batch, dec_state=None, enc_state=None):
         #tgt_lengths = batch.tgt_lengths
 
         # running forward() method in the NegotiationModel
-        if hasattr(self.model, 'context_embedder'):
+        if hasattr(batch, 'context_inputs'):
```
Comment (hhexiy) [inline cocoa/neural/trainer.py]: override run_seq2seq_batch function in fb trainer, remove from cocoa

File: cocoa/neural/trainer.py [hunk 9]
```diff
@@ -353,7 +369,11 @@ def _gradient_accumulation(self, true_batchs, total_stats, report_stats):
             self.model.zero_grad()
             outputs, attns, dec_state = self._run_batch(batch, None, enc_state)
 
-            loss, batch_stats = self.train_loss.compute_loss(batch.targets, outputs)
+            if hasattr(batch, "selections"):
+                loss, batch_stats = self.train_loss.compute_loss(batch.targets, batch.selections, outputs)
+            else:
+                loss, batch_stats = self.train_loss.compute_loss(batch.targets, outputs)
+
```
Comment (hhexiy) [inline cocoa/neural/trainer.py]: same. clean up trainer.

File: fb-negotiation/neural/model_builder.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+from models import FBNegotiationModel, LM
+
+from cocoa.io.utils import read_pickle
+from cocoa.pt_model.util import use_gpu
+
+from symbols import markers
+from neural import make_model_mappings
+
+def add_model_arguments(parser):
+    from onmt.modules.SRU import CheckSRU
```
Comment (hhexiy) [inline fb-negotiation/neural/model_builder.py:25]: import existing arguments from cocoa.neural.model_builder. only add those specific to fb here.

File: fb-negotiation/neural/model_builder.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+    word_padding_idx = word_dict.to_ind(markers.PAD)
+    num_word_embeddings = len(word_dict)
+
+    return Embeddings(word_vec_size=embedding_dim,
+                      position_encoding=False,
+                      dropout=opt.dropout,
+                      word_padding_idx=word_padding_idx,
+                      word_vocab_size=num_word_embeddings)
+
```
Comment (hhexiy) [inline fb-negotiation/neural/model_builder.py:116]: can you import this from cocoa

File: fb-negotiation/neural/model_builder.py [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+                          opt.cnn_kernel_width,
+                          opt.dropout, embeddings)
+    elif opt.encoder_type == "mean":
+        return MeanEncoder(opt.enc_layers, embeddings)
+    else:
+        # "rnn" or "brnn"
+        bidirectional = True if opt.encoder_type == 'brnn' else False
+        return StdRNNEncoder(opt.rnn_type, bidirectional, opt.enc_layers,
+                        opt.rnn_size, opt.dropout, embeddings)
+
```
Comment (hhexiy) [inline fb-negotiation/neural/model_builder.py:139]: same, import from cocoa

File: fb-negotiation/neural/model_builder.py [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+        "my selection", selectors 4 to 6 are for "partner's selection".
+    '''
+    selectors["dec"] = nn.ModuleList()
+    for i in xrange(6):
+        selectors["dec"].append(nn.Linear(opt.sel_hid_size, kb_dict.size))
+
+    return selectors
+
+def make_context_embedder(opt, embeddings, embed_type='utterance'):
+    """
```
Comment (hhexiy) [inline fb-negotiation/neural/model_builder.py]: import from cocoa

File: fb-negotiation/neural/model_builder.py [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+    if opt.context_embedder_type == "mean":
+        return MeanEncoder(opt.enc_layers, embeddings, embed_type)
+    else:
+        # "rnn" or "brnn"
+        bidirectional = True if opt.context_embedder_type == 'brnn' else False
+        return StdRNNEncoder(opt.rnn_type, bidirectional, opt.enc_layers,
+                        opt.rnn_size, opt.dropout, embeddings, embed_type, False)
+
+def make_decoder(opt, embeddings, tgt_dict):
+    """
```
Comment (hhexiy) [inline fb-negotiation/neural/model_builder.py:191]: import form cocoa

Comment (hhexiy) [reply fb-negotiation/neural/model_builder.py:191]: this one you need some override. but you can always call the cocoa one first, and add whatever you need after. it.

File: fb-negotiation/neural/model_builder.py [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+    else:
+        return StdRNNDecoder(opt.rnn_type, bidirectional,
+                             opt.dec_layers, opt.rnn_size,
+                             attn_type=opt.global_attention,
+                             dropout=opt.dropout,
+                             embeddings=embeddings,
+                             pad=pad)
+
+def load_test_model(model_path, opt, dummy_opt):
+    checkpoint = torch.load(model_path, map_location=lambda storage, loc: storage)
```
Comment (hhexiy) [inline fb-negotiation/neural/model_builder.py:230]: import from cocoa

File: fb-negotiation/neural/models.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
+import torch
+import torch.nn as nn
+
+from cocoa.neural.models import NMTModel
+
+class LM(nn.Module):
+    def __init__(self, encoder):
```
Comment (hhexiy) [inline fb-negotiation/neural/models.py]: remove LM

File: fb-negotiation/neural/symbols.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+    # Actions
+    SELECT = '<select>'
+    # OFFER = '<offer>'
+    # ACCEPT = '<accept>'
+    # REJECT = '<reject>'
+    QUIT = '<quit>'
+
+    # Items
+    C_book = '<book>'
```
Comment (hhexiy) [inline fb-negotiation/neural/symbols.py]: remove these

File: fb-negotiation/neural/generator.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        decoder_outputs, dec_states, _ = self.model.decoder(
+            inp, memory_bank, dec_states, memory_lengths=lengths)
+
+        # (2) Sampling
+        batch_size = batch.size
+        preds = []
+        for i in xrange(self.max_length):
+            # Outputs to probs
+            decoder_outputs = decoder_outputs.squeeze(0)  # (batch_size, rnn_size)
+            out = self.model.generator.forward(decoder_outputs).data  # Logprob (batch_size, vocab_size)
```
Comment (hhexiy) [inline fb-negotiation/neural/generator.py:34]: does this run?? isn't decoder_outputs a dict?

File: cocoa/core/dataset.py (hunk 1 above)
Comment (derekchen14) [reply cocoa/core/dataset.py]: changed to match the unit_test below, i guess the better idea would have been to change the unit_test to match the actual code.  i will do that now

File: cocoa/neural/evaluator.py (hunk 2 above)
Comment (derekchen14) [reply cocoa/neural/evaluator.py:80]: the counter is for knowing the example number, and for extracting the response (which depends on the sentence number).  originally, when the evaluate() and print_results() were one function the counter did not need to be passed around, but now it does.  we could create some more global counter, but that seems messy as well

File: cocoa/neural/models.py (hunk 5 above)
Comment (derekchen14) [reply cocoa/neural/models.py]: num_layers is specified by command line, and i change it ad-hoc for scenario which expects (1 x batch_size x hidden_dim), rather than (num_layers x batch_size x hidden_dim).  i agree this is messy.  how would you handle?  I guess that depends on how we end up wiring together the model.forward() function?

File: cocoa/neural/models.py (hunk 6 above)
Comment (derekchen14) [reply cocoa/neural/models.py]: Originally, encoder_final was a tuple of (mean, mean) which seemed redundant, but looking back, it probably has some purpose.

File: fb-negotiation/neural/symbols.py (hunk 17 above)
Comment (derekchen14) [reply fb-negotiation/neural/symbols.py]: removed the "item" markers

File: fb-negotiation/neural/generator.py (hunk 18 above)
Comment (derekchen14) [reply fb-negotiation/neural/generator.py:34]: yes it runs. i think decoder_outputs is a Variable, but this difference in thought seems critical.  Perhaps decoder_outputs *should* be a dict and somewhere it was changed to Variable incorrectly.

File: cocoa/neural/models.py (hunk 5 above)
Comment (hhexiy) [reply cocoa/neural/models.py]: you can add a separate cmd arguments just for the embedder

File: cocoa/neural/models.py (hunk 6 above)
Comment (hhexiy) [reply cocoa/neural/models.py]: well... we should not change their interface... one is state the other is output. if you're not having problems with it just keep it the way it is.

Comment (hhexiy) [reply cocoa/neural/models.py]: in general unless you're 100% sure what the code is doing don't change it.

File: fb-negotiation/neural/generator.py (hunk 18 above)
Comment (hhexiy) [reply fb-negotiation/neural/generator.py:34]: well then it should not be running right? your FBNegotiationModel returns outputs as a dict.....
